File: src/main/java/thedarkcolour/core/block/BlockBase.java
package thedarkcolour.core.block;

import net.minecraft.block.Block;
import net.minecraft.block.SoundType;
import net.minecraft.block.material.MapColor;
import net.minecraft.block.material.Material;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.util.math.AxisAlignedBB;
import thedarkcolour.futuremc.FutureMC;

public class BlockBase extends Block {
    public BlockBase(String regName) {
        this(regName, Material.ROCK);
    }

    public BlockBase(String regName, Material material) {
        this(regName, material, SoundType.STONE);
    }

    public BlockBase(String regName, Material material, SoundType soundType) {
        super(material);
        setTranslationKey(FutureMC.ID + "." + regName);
        setRegistryName(regName);
        setSoundType(soundType);
        setHardness(3.0F);
    }

    public BlockBase(String regName, Material material, MapColor color, SoundType soundType) {
        super(material, color);
        setTranslationKey(FutureMC.ID + "." + regName);
        setRegistryName(regName);
        setSoundType(soundType);
        setHardness(3.0F);
    }

    public static AxisAlignedBB makeAABB(double startX, double startY, double startZ, double endX, double endY, double endZ) {
        return new AxisAlignedBB(startX / 16D, startY / 16D, startZ / 16D, endX / 16D, endY / 16D, endZ / 16D);
    }

    @Override
    public BlockBase setCreativeTab(CreativeTabs tab) {
        super.setCreativeTab(tab);
        return this;
    }

    @Override
    public BlockBase setHardness(float hardness) {
        super.setHardness(hardness);
        return this;
    }
}

File: src/main/java/thedarkcolour/core/block/BlockBase.kt
package thedarkcolour.core.block

import net.minecraft.block.Block
import net.minecraft.block.SoundType
import net.minecraft.block.material.MapColor
import net.minecraft.block.material.Material
import net.minecraft.item.EnumDyeColor
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.AxisAlignedBB
import thedarkcolour.futuremc.FutureMC

open class BlockBase : Block {
    @JvmOverloads
    @Deprecated("Use props constructor", ReplaceWith("()as"))
    constructor(regName: String, material: Material = Material.ROCK, soundType: SoundType = SoundType.STONE)
            : super(material) {
        translationKey = "${FutureMC.ID}.$regName"
        setRegistryName(regName)
        blockSoundType = soundType
        blockHardness = 3.0f
    }

    @Deprecated("Use props constructor", ReplaceWith("()"))
    constructor(regName: String, material: Material, color: MapColor, soundType: SoundType) : super(material, color) {
        translationKey = "${FutureMC.ID}.$regName"
        setRegistryName(regName)
        blockSoundType = soundType
        blockHardness = 3.0f
    }

    constructor(properties: Props) : super(
        properties.material,
        properties.color ?: properties.material.materialMapColor
    ) {
        translationKey = FutureMC.ID + "." + properties.registryName
        registryName = ResourceLocation(FutureMC.ID, properties.registryName)
        blockSoundType = properties.sound
        blockHardness = properties.hardness
    }

    companion object {
        fun makeAABB(
            startX: Double,
            startY: Double,
            startZ: Double,
            endX: Double,
            endY: Double,
            endZ: Double
        ): AxisAlignedBB {
            return AxisAlignedBB(startX / 16.0, startY / 16.0, startZ / 16.0, endX / 16.0, endY / 16.0, endZ / 16.0)
        }
    }

    class Props(val registryName: String) {
        var material = Material.ROCK
        var sound = SoundType.STONE
        var color: MapColor? = null
        var hardness = 3f

        fun material(material: Material) {
            this.material = material
        }

        fun color(color: MapColor) {
            this.color = color
        }

        fun color(color: EnumDyeColor) {
            this.color = MapColor.getBlockColor(color)
        }

        fun sound(sound: SoundType) {
            this.sound = sound
        }

        fun hardness(hardness: Float) {
            this.hardness = hardness
        }
    }
}